File: Shared/MTAFeedConfiguration.swift
//
//  MTAFeedConfiguration.swift
//  Now Departing
//
//  MTA GTFS-RT feed URL configuration and route-to-feed mapping.
//  No API key is required for subway realtime feeds.
//

import Foundation

struct MTAFeedConfiguration {

    // MARK: - Base URL

    static let baseURL = "https://api-endpoint.mta.info/Dataservice/mtagtfsfeeds/nyct"

    // MARK: - Feed Paths
    //
    // Multiple routes share a single feed file; fetching any route from the
    // group retrieves real-time data for every route in that group.
    //
    // Feed groups:
    //   gtfs      → 1 2 3 4 5 6 7 S (numbered lines + 42nd St Shuttle)
    //   gtfs-ace  → A C E FS (Franklin Ave Shuttle) H (Rockaway Park)
    //   gtfs-bdfm → B D F M
    //   gtfs-g    → G
    //   gtfs-jz   → J Z
    //   gtfs-nqrw → N Q R W
    //   gtfs-l    → L
    //   gtfs-si   → SIR (Staten Island Railway)

    static let feedPathByRoute: [String: String] = [
        "1":  "gtfs",
        "2":  "gtfs",
        "3":  "gtfs",
        "4":  "gtfs",
        "5":  "gtfs",
        "6":  "gtfs",
        "GS": "gtfs",
        "A":  "gtfs-ace",
        "C":  "gtfs-ace",
        "E":  "gtfs-ace",
        "FS": "gtfs-ace",
        "H":  "gtfs-ace",
        "B":  "gtfs-bdfm",
        "D":  "gtfs-bdfm",
        "F":  "gtfs-bdfm",
        "M":  "gtfs-bdfm",
        "G":  "gtfs-g",
        "J":  "gtfs-jz",
        "Z":  "gtfs-jz",
        "N":  "gtfs-nqrw",
        "Q":  "gtfs-nqrw",
        "R":  "gtfs-nqrw",
        "W":  "gtfs-nqrw",
        "L":  "gtfs-l",
        "7":  "gtfs",
        "SI": "gtfs-si",
    ]

    // MARK: - URL Helpers

    /// Returns the GTFS-RT feed URL for a given route ID, or nil if unknown.
    static func feedURL(for routeId: String) -> URL? {
        guard let path = feedPathByRoute[routeId] else { return nil }
        return URL(string: "\(baseURL)/\(path)")
    }

    /// Returns the deduplicated list of feed URLs for a set of route IDs.
    /// Routes that share a feed file are collapsed to a single URL.
    static func feedURLs(for routeIds: [String]) -> [URL] {
        var seenPaths = Set<String>()
        var urls: [URL] = []
        for route in routeIds {
            guard let path = feedPathByRoute[route], !seenPaths.contains(path) else { continue }
            seenPaths.insert(path)
            if let url = URL(string: "\(baseURL)/\(path)") {
                urls.append(url)
            }
        }
        return urls
    }

    // MARK: - Request Builder

    /// Builds a URLRequest for a feed URL.
    static func request(for url: URL) -> URLRequest {
        var req = URLRequest(url: url)
        req.timeoutInterval = 30
        return req
    }
}

File: NowDepartingAndroid/app/src/main/java/com/move38/nowdeparting/data/api/MTAFeedConfiguration.kt
package com.move38.nowdeparting.data.api

object MTAFeedConfiguration {

    private const val BASE_URL = "https://api-endpoint.mta.info/Dataservice/mtagtfsfeeds/nyct%2F"

    // Multiple routes share a single feed file; fetching any route from the
    // group retrieves real-time data for every route in that group.
    //
    // Feed groups:
    //   gtfs      → 1 2 3 4 5 6 7 GS (numbered lines + 42nd St Shuttle)
    //   gtfs-ace  → A C E FS (Franklin Ave Shuttle) H (Rockaway Park)
    //   gtfs-bdfm → B D F M
    //   gtfs-g    → G
    //   gtfs-jz   → J Z
    //   gtfs-nqrw → N Q R W
    //   gtfs-l    → L
    //   gtfs-si   → SI (Staten Island Railway)

    private val feedPathByRoute = mapOf(
        "1"  to "gtfs",
        "2"  to "gtfs",
        "3"  to "gtfs",
        "4"  to "gtfs",
        "5"  to "gtfs",
        "6"  to "gtfs",
        "7"  to "gtfs",
        "GS" to "gtfs",
        "A"  to "gtfs-ace",
        "C"  to "gtfs-ace",
        "E"  to "gtfs-ace",
        "FS" to "gtfs-ace",
        "H"  to "gtfs-ace",
        "B"  to "gtfs-bdfm",
        "D"  to "gtfs-bdfm",
        "F"  to "gtfs-bdfm",
        "M"  to "gtfs-bdfm",
        "G"  to "gtfs-g",
        "J"  to "gtfs-jz",
        "Z"  to "gtfs-jz",
        "N"  to "gtfs-nqrw",
        "Q"  to "gtfs-nqrw",
        "R"  to "gtfs-nqrw",
        "W"  to "gtfs-nqrw",
        "L"  to "gtfs-l",
        "SI" to "gtfs-si"
    )

    /** GTFS-RT service alerts feed URL (all subway lines). */
    val alertsFeedUrl: String = "${BASE_URL}gtfs-alerts"

    /** Returns the GTFS-RT feed URL for a given route ID, or null if unknown. */
    fun feedUrl(routeId: String): String? {
        val path = feedPathByRoute[routeId] ?: return null
        return "$BASE_URL$path"
    }

    /** Returns deduplicated feed URLs for a set of route IDs. Routes that share
     *  a feed file are collapsed to a single URL. */
    fun feedUrls(routeIds: List<String>): List<String> {
        val seenPaths = mutableSetOf<String>()
        val urls = mutableListOf<String>()
        for (routeId in routeIds) {
            val path = feedPathByRoute[routeId] ?: continue
            if (seenPaths.add(path)) {
                urls.add("$BASE_URL$path")
            }
        }
        return urls
    }
}
